Enable range index on raw columns and push counts down to range index when v2 is in use

richardstartin opened the pull request:
This extends the fast count mechanism to v2 range indexes now that the underlying data structure supports counting. Also does some code cleanup in this area.

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:
```diff
@@ -145,10 +127,178 @@ public List<Operator> getChildOperators() {
+    IntRangeEvaluator(
+        RangeIndexReader<ImmutableRoaringBitmap> rangeIndexReader, PredicateEvaluator predicateEvaluator) {
+      _rangeIndexReader = rangeIndexReader;
+      if (predicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator) {
+        // NOTE: End dictionary id is exclusive in OfflineDictionaryBasedRangePredicateEvaluator.
+        _min = ((SortedDictionaryBasedRangePredicateEvaluator) predicateEvaluator).getStartDictId();
+        _max = ((SortedDictionaryBasedRangePredicateEvaluator) predicateEvaluator).getEndDictId() - 1;
```
why is the DictionaryBasedRangePredicateEvaluator branch only exist in the IntRangeEvaluate?

richardstartin replied:
the purpose of this is to abstract over the range bound types - int, long, etc. Only raw int and dictionary have int bounds.

walterddr commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:
```diff
@@ -26,6 +26,51 @@
+   * This means {@see getPartiallyMatchingDocIds} will return null.
+   */
+  default boolean isExact() {
+    return true;
```
I would not make this a default. shouldn't the implementer decide?

richardstartin replied:
There are two implementations, the first one was inexact and could be up to 10x slower than the one which is exact. If we implement another range index, we will want it to be exact.

walterddr replied:
oh yes. what I meant is we should not implement a default --> whether the implementation is exact or not cannot be determined in the interface level - e.g. if I accidentally deleted the override in `RangeIndexReaderImpl` i will get no compile time error until i run into test failures. 

is the intent here to not allow any non-exact impl in the future?

richardstartin replied:
> is the intent here to not allow any non-exact impl in the future?

exactly, you need to opt out of being exact and have a good reason

walterddr reviewed (commented):
looks good to me overall. I just have 2 questions. please kindly take a look

Jackie-Jiang commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:
```diff
@@ -26,6 +26,51 @@
+   */
+  default boolean isExact() {
+    return true;
+  }
```
(minor) Add an empty line after

Jackie-Jiang commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:
```diff
@@ -26,6 +26,51 @@
+   * @param max the inclusive upper bound.
+   * @return the matching doc ids.
+   */
+  int getMatchingDocCount(int min, int max);
```
(minor) Suggest renaming to `getNumMatchingDocs` to be consistent with `BaseFilterOperator`

Jackie-Jiang commented on pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/BitSlicedIndexCreatorTest.java:
```diff
@@ -192,7 +194,9 @@ private void testLong(Dataset<long[]> dataset)
         ImmutableRoaringBitmap result = reader.getMatchingDocIds(prev, quantile);
+        int resultCount = reader.getMatchingDocCount(prev, quantile);
         assertEquals(result, reference);
+        assertEquals(resultCount, result == null ? 0 : result.getCardinality());
```
Per the implementation seems `result` can never be `null`, and we rely on it not being `null` in `RangeIndexBasedFilterOperator.getNextBlock()` or it will throw NPE

richardstartin replied:
I haven't changed this, I would prefer the unnecessary code to the IDE warning.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:
```diff
@@ -145,10 +127,178 @@ public List<Operator> getChildOperators() {
+
+    ImmutableRoaringBitmap getPartiallyMatchingDocIds();
+
+    int getMatchingDocCount();
```
(minor) Suggest renaming to `getNumMatchingDocs` to be consistent with `BaseFilterOperator`

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:
```diff
@@ -145,10 +127,178 @@ public List<Operator> getChildOperators() {
+    DoubleRangeEvaluator(RangeIndexReader<ImmutableRoaringBitmap> rangeIndexReader,
+        PredicateEvaluator predicateEvaluator) {
+      _rangeIndexReader = rangeIndexReader;
+      _min = ((FloatRawValueBasedRangePredicateEvaluator) predicateEvaluator).geLowerBound();
```
(MAJOR) This should be `DoubleRawValueBasedRangePredicateEvaluator`. We should probably add a test which can catch this

richardstartin replied:
Oh wow, thanks for spotting this! I was sure this code path was tested. I will find out why not.

richardstartin replied:
Fixed.

Jackie-Jiang reviewed (approved):
LGTM otherwise. There is a major comment on a copy-pasted bug

richardstartin commented:
@Jackie-Jiang the copy-paste bug you spotted was revealing - range indexes over raw columns was not actually supported, despite #5853 aiming to add support. Commit fe28919f0dd9b0da143263266f4bba2ddab5b3e0 addresses that. Commit f30099e9b654fc2d92aba20afd72722b3a1003dc addresses your review comments, and adds a test to verify the bug is fixed (and that the code executes).
